Tests/AsyncHTTPClientTests/ConnectionPoolTests.swift

```diff
@@ -0,0 +1,1437 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2018-2019 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+@testable import AsyncHTTPClient
+@testable import NIO
+import NIOConcurrencyHelpers
+import NIOFoundationCompat
+import NIOHTTP1
+import NIOHTTPCompression
+import NIOSSL
+import NIOTestUtils
+import NIOTransportServices
+import XCTest
+
+class ConnectionPoolTests: XCTestCase {
+    struct TempError: Error {}
+
+    func testPending() {
+        let eventLoop = EmbeddedEventLoop()
+
+        var state = HTTP1ConnectionProvider.ConnectionsState(eventLoop: eventLoop)
+
+        var snapshot = state.testsOnly_getInternalState()
+        XCTAssertEqual(0, snapshot.availableConnections.count)
```

Comment:
they look all great! Quick question: Why are we asserting on the (private) state and not the `Action`s that the state machine returns? Examples here: https://github.com/apple/swift-nio-examples/blob/master/backpressure-file-io-channel/Tests/BackpressureChannelToFileIOTests/StateMachineTest.swift#L267-L309

For the example above, I extended the `Action` to have lots of functions (one for each of the actions) like

```
    @discardableResult
    func assertOpenFile(_ check: (String) throws -> Void,
                        file: StaticString = #file,
                        line: UInt = #line) -> Self {
        if case .openFile(let path) = self.main {
            XCTAssertNoThrow(try check(path))
        } else {
            XCTFail("action \(self) not \(#function)", file: file, line: line)
        }
        return self
    }
```

And then you can just write

```
self.stateMachine.someInput()
    .assertActionIsX()
    .assertWhateverElse()
```

and we wouldn't need to expose the state. But again, I think that's something we can do in the future.

The benefit of asserting on the `Action` is that we can change the state machine representation to say an `enum` without rewriting all the tests.

Reply:
we do assert on actions as well (not in this test specifically, but in other tests), I want to test not only action but the way state is changing. I test before the action to "show" what state I was expecting in this test and after to see if we arrived at it, also serves as a documentation of sorts for the test case

Reply:
Hmm, but the tests shouldn't care how the state machine achieves its goal, assuming it does, right? When we refactor, we'll lose all the 'private state tests'.
The state machine is fully deterministic, so there isn't anything you can't test with just inputs & actions. The state is fully determined by the inputs that have been fed.

As long as most of the tests do only action assertions, I think there's no harm caused also asserting on private state but I think it's a bit of an anti-pattern.

Reply:
I agree, but it can be a bit easier to test that when we `aqcuire` a connection that it is inside the waiters for example, than to get a series of transformations to `aquire` and then release some other connection and then test that our promise was succeeded

Reply:
Github’s broken, i was meaning to thumbs UP only but fat fingered a DOWN too which I can’t remove now 🤪